src/API/Google/ExceptionTrait.php

```diff
@@ -82,6 +114,23 @@ protected function get_api_exception_errors( ApiException $exception ): array {
 		return $errors;
 	}
 
+	/**
+	 * Returns a list of detailed errors from a GoogleServiceException.
+	 *
+	 * @param GoogleServiceException $exception Exception to check.
+	 *
+	 * @return array
+	 */
+	private function get_google_service_exception_errors( GoogleServiceException $exception ): array {
+		$errors = [];
+
+		foreach ( $exception->getErrors() as $error ) {
+			$errors[ $error['reason'] ] = $error['message'];
```

Comment:
How we can be sure `$error['reason'` and `$error['message']` are defined?  

I would add some kind of [isset](https://www.php.net/manual/en/function.isset.php) to be sure is defined, and return a generic otherwise.

Reply:
From Google's source code and doc, I think $error['reason']` and `$error['message']` should always exist:

- Only [`Google\Http\REST::decodeHttpResponse`](https://github.com/googleapis/google-api-php-client/blob/v2.15.0/src/Http/REST.php#L121-L135) may create `GoogleServiceException` instances along with calling [`getResponseErrors`](https://github.com/googleapis/google-api-php-client/blob/v2.15.0/src/Http/REST.php#L176-L185)
- [`GoogleServiceException::getErrors`](https://github.com/googleapis/google-api-php-client/blob/v2.15.0/src/Service/Exception.php#L54-L72) provides an example of the returned errors 
- [Google doc](https://developers.google.com/shopping-content/reference/rest/v2.1/Errors) shows the errors returned by Google Shopping Content service

Do you think it's still better to add `isset` checks?

BTW, while tracing the above things, I noticed that `$exception->getErrors()` might return `null`, so I added the null check in https://github.com/woocommerce/google-listings-and-ads/pull/2007/commits/8e46c2ae48e2e9d383fa86c5a6073ce7b637ef98.

Reply:
I'm always into checking 3rd party inputs... even if it's from Google and it's documented.  Even if actually is always defined, what will happen if we update the GLA library or the API and they change this? Seems to me that It's going to be missed since we are not going to test the error messages as frequently. 

Since it's not a big change and it's definitely not harmful to the code I would suggest adding it. What do you think?

Reply:
> I'm always into checking 3rd party inputs... even if it's from Google and it's documented. Even if actually is always defined, what will happen if we update the GLA library or the API and they change this? Seems to me that It's going to be missed since we are not going to test the error messages as frequently.

I'm hard to agree that it's reasonable to "**always** do data structure checking to 3rd-party inputs", rather it's a bit overkill.

Fault tolerance is usually used to allow the application to recover from errors such as retry when the network is unstable, or not to result in a critical problem like system crash.

However, when the data structure provided by a 3rd-party is suddenly not presented in its promised format, the application usually doesn't work, and even fault tolerance for missing or breaking changed data would hardly change the result of the unavailable application.

> Since it's not a big change and it's definitely not harmful to the code I would suggest adding it.

Although I don't think it's necessary, I also don't strongly disagree it. Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/2007/commits/8fd35cc10b76ce1a8b7561cf5ffb3239a4a85644.